### PR TITLE
[PT DateTimeV2] Support datetime like "da tarde", "ao meio dia", "da noite"

### PR DESCRIPTION
**Pull request by ChopperMan33**
Support datetime like "da tarde", "ao meio dia", "da noite" in PT.
Add test cases.

### CODE REVIEW

File: Specs/DateTime/Portuguese/DateTimeModel.json [hunk 1]
```diff
@@ -1328,5 +1328,255 @@
...
+              "type": "time",
+              "value": "16:00:00"
+            }
+          ]
+        }
+      }
+    ]
+  },
+  {
+    "Input": "Lembre-se da noite do dia 21 de setembro",
```
Comment (tellarin) [inline Specs/DateTime/Portuguese/DateTimeModel.json:1552]: Does this PR also already solves #2481?

**Review by tellarin — CHANGES_REQUESTED** [review]

File: .NET/Microsoft.Recognizers.Text.DateTime/Portuguese/Parsers/DateTimePeriodParser.cs [hunk 2]
```diff
@@ -55,7 +55,16 @@ protected override DateTimeResolutionResult ParseSpecificTimeOfDay(string text,
 
                 if (ers.Count == 0)
                 {
-                    return ret;
+                    // prefix match
+                    var afterStr = trimmedText.Substring(match.Index + startIndex + match.Length).Trim();
```
[the remark this answers is not in the document]
Comment (ChopperMan33) [reply .NET/Microsoft.Recognizers.Text.DateTime/Portuguese/Parsers/DateTimePeriodParser.cs:59]: This is to deal with the case of DataTime before Period.

File: .NET/Microsoft.Recognizers.Text.DateTime/Portuguese/Parsers/DateTimePeriodParser.cs [hunk 3]
```diff
@@ -55,7 +55,16 @@ protected override DateTimeResolutionResult ParseSpecificTimeOfDay(string text,
...
+                    ers = this.Config.DateExtractor.Extract(afterStr, referenceTime);
+
+                    if (ers.Count == 0)
+                    {
+                        return ret;
+                    }
+
+                    beforeStr = afterStr;
```
[the remark this answers is not in the document]
Comment (ChopperMan33) [reply .NET/Microsoft.Recognizers.Text.DateTime/Portuguese/Parsers/DateTimePeriodParser.cs:67]: The same as before, mainly to solve the problem of string division, the code logic behind is consistent with the "after" situation.

File: .NET/Microsoft.Recognizers.Text.DateTime/Portuguese/Parsers/PortugueseDateTimePeriodParserConfiguration.cs [hunk 4]
```diff
@@ -134,34 +135,64 @@ public bool GetMatchedTimeRange(string text, out string timeStr, out int beginHo
...
+                beginHour = Constants.HalfDayHourCount;
+                endHour = 16;
+            }
+            else if (DateTimeDefinitions.EveningTermList.Any(o => trimmedText.EndsWith(o, StringComparison.Ordinal)))
+            {
+                timeStr = Constants.Evening;
+                beginHour = 16;
+                endHour = 20;
+            }
+            else if (DateTimeDefinitions.NightTermList.Any(o => trimmedText.EndsWith(o, StringComparison.Ordinal)))
```
[the remark this answers is not in the document]
Comment (ChopperMan33) [reply .NET/Microsoft.Recognizers.Text.DateTime/Portuguese/Parsers/PortugueseDateTimePeriodParserConfiguration.cs:162]: Solve the problem of DateTime behind Period, and for some cases, keep priority matching DateTime before Period.

File: Specs/DateTime/Portuguese/DateTimeModel.json [hunk 5]
```diff
@@ -1328,5 +1328,255 @@
...
+  },
+  {
+    "Input": "meio da tarde",
+    "Context": {
+      "ReferenceDateTime": "2021-01-15T18:00:00"
+    },
+    "NotSupported": "javascript,python",
+    "Results": [
+      {
+        "Text": "meio da tarde",
```
[the remark this answers is not in the document]
Comment (ChopperMan33) [reply Specs/DateTime/Portuguese/DateTimeModel.json:1438]: Does "meio da tarde" mean "mid-afternoon"? Should we return "afternoon" and "mid" separately?

File: Specs/DateTime/Portuguese/TimeExtractor.json [hunk 6]
```diff
@@ -568,10 +568,10 @@
     "NotSupported": "javascript, python",
     "Results": [
       {
-        "Text": "de madrugada",
+        "Text": "madrugada",
```
[the remark this answers is not in the document]
Comment (ChopperMan33) [reply Specs/DateTime/Portuguese/TimeExtractor.json:571]: So, should we return an empty list in this case?

File: .NET/Microsoft.Recognizers.Text.DateTime/Portuguese/Parsers/DateTimePeriodParser.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ChopperMan33) [reply .NET/Microsoft.Recognizers.Text.DateTime/Portuguese/Parsers/DateTimePeriodParser.cs:59]: Like 
```json
 {
    "Input": "Lembre-se da noite do dia 21 de setembro",
    "Context": {
      "ReferenceDateTime": "2021-01-15T18:00:00"
    },
    "NotSupported": "javascript,python",
    "Results": [
      {
        "Text": "noite do dia 21 de setembro",
        "Start": 13,
        "End": 39,
        "TypeName": "datetimeV2.datetimerange",
        "Resolution": {
          "values": [
            {
              "timex": "XXXX-09-21TNI",
              "type": "datetimerange",
              "start": "2020-09-21 20:00:00",
              "end": "2020-09-21 23:59:59"
            },
            {
              "timex": "XXXX-09-21TNI",
              "type": "datetimerange",
              "start": "2021-09-21 20:00:00",
              "end": "2021-09-21 23:59:59"
            }
          ]
        }
      }
    ]
  }
```

File: Specs/DateTime/Portuguese/DateTimeModel.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (ChopperMan33) [reply Specs/DateTime/Portuguese/DateTimeModel.json:1438]: I refer the engish version.

```json
{
  "Text": "midafternoon",
  "Start": 0,
  "End": 11,
  "TypeName": "datetimeV2.time",
  "Resolution": {
    "values": [
      {
        "timex": "T14",
        "type": "time",
        "value": "14:00:00"
      }
    ]
  }
}
```

Do we need to modify to timerange

File: Specs/DateTime/Portuguese/TimeExtractor.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (ChopperMan33) [reply Specs/DateTime/Portuguese/TimeExtractor.json:571]: Should this case be, due to it's time extractor file.
```json
  {
    "Input": "Voltarei de madrugada",
    "NotSupported": "javascript, python",
    "Results": []
  },
```

File: .NET/Microsoft.Recognizers.Definitions.Common/Portuguese/DateTimeDefinitions.cs [hunk 7]
```diff
@@ -133,13 +138,13 @@ public static class DateTimeDefinitions
...
+      public static readonly string TimeRegex10 = $@"{TimeTokenPrefixRegex}(madrugada|manh[ãa])";
       public static readonly string TimeRegex11 = $@"\b({WrittenTimeRegex})(\s+{DescRegex})?\b";
       public static readonly string TimeRegex12 = $@"(\b{TimePrefix}\s+)?{BaseDateTime.HourRegex}(\s*h\s*){BaseDateTime.MinuteRegex}(\s*{DescRegex})?";
       public const string PrepositionRegex = @"(?<prep>([àa]s?|em|por|pelo|pela|no|na|de|d[oa]?)?$)";
       public const string NowRegex = @"\b(?<now>((logo|exatamente)\s+)?agora(\s+mesmo)?|neste\s+momento|(assim\s+que|t[ãa]o\s+cedo\s+quanto)\s+(poss[ií]vel|possas?|possamos)|o\s+mais\s+(cedo|r[aá]pido)\s+poss[íi]vel|recentemente|previamente)\b";
       public const string SuffixRegex = @"^\s*((e|a|em|por|pelo|pela|no|na|de)\s+)?(manh[ãa]|madrugada|meio\s*dia|tarde|noite)\b";
-      public const string TimeOfDayRegex = @"\b(?<timeOfDay>manh[ãa]|madrugada|tarde|noite|((depois\s+do|ap[óo]s\s+o)\s+(almo[çc]o|meio dia|meio-dia)))\b";
+      public const string TimeOfDayRegex = @"\b(?<timeOfDay>manh[ãa]|madrugada|tarde|noite|no\s+meio\s+do\s+dia|((depois\s+do|ap[óo]s\s+o)\s+(almo[çc]o|meio dia|meio-dia)))\b";
```
[the remark this answers is not in the document]
Comment (ChopperMan33) [reply .NET/Microsoft.Recognizers.Definitions.Common/Portuguese/DateTimeDefinitions.cs:147]: "in the middle of the day " is not supported in English, only support "midday" as T12

File: .NET/Microsoft.Recognizers.Text.DateTime/Portuguese/Parsers/PortugueseDateTimePeriodParserConfiguration.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ChopperMan33) [reply .NET/Microsoft.Recognizers.Text.DateTime/Portuguese/Parsers/PortugueseDateTimePeriodParserConfiguration.cs:162]: Yes, I will simplify it.

File: Specs/DateTime/Portuguese/DateTimeModel.json (hunk 5 above)
[the remark this answers is not in the document]
Comment (tellarin) [reply Specs/DateTime/Portuguese/DateTimeModel.json:1438]: Yeah... this is clearly incorrect.

File: Specs/DateTime/Portuguese/TimeExtractor.json (hunk 6 above)
[the remark this answers is not in the document]
Comment (tellarin) [reply Specs/DateTime/Portuguese/TimeExtractor.json:571]: Same as "tarde".

**Comment by tellarin** [comment]
Closing stale PR.

**Pull request closed without merging**